contribute/design/1.x/cstor/20200629-cstor-backupandrestore-v1.md

```diff
@@ -0,0 +1,278 @@
+	Size         resource.Quantity `json:"size,omitempty"`
+	// Local defines whether restore is from local/remote
+	Local bool `json:"localRestore,omitempty"` // if restore is from local backup/snapshot
+}
+
+// CStorRestoreStatus is to hold result of action.
+type CStorRestoreStatus string
+```
+
+Following is the proposed CStorCompletedBackup schema in go struct:
+```go
+// CStorCompletedBackup describes a cstor completed-backup resource created as custom resource
+type CStorCompletedBackup struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+    Spec              CStorCompletedBackupSpec `json:"spec"`
+}
+
+// CStorCompletedBackupSpec is the spec for a CStorBackup resource
+type CStorCompletedBackupSpec struct {
+	// BackupName is a name of the backup or scheduled backup
+	BackupName string `json:"backupName"`
+
+	// VolumeName is a name of the volume for which this backup is destined
+	VolumeName string `json:"volumeName"`
+
+	// SecondLastSnapName is a name of the second last completed-backup's snapshot
+	SecondLastSnapName string `json:"secondLastSnapName"`
+
+	// LastSnapName is the last completed-backup's snapshot name
+	LastSnapName string `json:"lastSnapName"`
+}
+```
+
+From existing to proposed has only one schema change i.e CStorCompletedBackupSpec has been introduced. Previously spec of CStorCompletedBackup is same as CStorBackup Spec. Since spec of CStorBackup doesn't fit for CStorCompletedBackup usability issues has been raised. To incorporate and decouple specs of both resources a new spec i.e `CStorCompletedBackupSpec` has been introduced. Now CStorCompletedBackup will hold CStorCompletedBackupSpec as a spec of it.
+
+##### Migration Workflows
+
+###### Approach1(With Backward compatibility)
+
+- Converting group and version of cStor backup and restore resources from `openebs.io/v1alpha1` to `cstor.openebs.io/v1` will break the support of backup and restore if control plane(CVC-Operator) alone is upgraded to latest version i.e 1.12.0 and data plane(cStor pools) is in lower version i.e 1.11.0. When control plane is upgraded it will understand `cstor.openebs.io` group but data plane is in old version so it will understand `openebs.io` group. To fix compatibility issues OpenEBS(CVC-Operator) needs to support backward compatibility for alpha features(backup and restore).
+- How can we fix backward compatibility? We can achieve it by following steps:
+  1. Whenever a CVC-Operator needs to perform CRUD request on backup and restore endpoints it will find the version of cStor pool by using payload(based on volume it will fetch CVR and using CVR labels CSPI info can be fetched) which was sent during request and takes a decision on which group it has to perform CRUD operations. For example if pool version is equal to 1.11.0 then CVC-Operator will make use `openebs.io` group client or if pool version is 1.12.0 then it will make use of `cstor.openebs.io` group.
```

Comment:
During upgrade We may need to check for in-progress backup/restore request, since fetching CVR and creating backup/restore resource is not atomic operation. WDYT?

Reply:
Updated the steps on how to handle the request during upgrade from one CSPC version to another CSPC version.